ct/l1: add compaction metadata

andrwng opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->
Adds the compaction metadata to the `metastore` interface, and an implementation in `simple_metastore` that is backed by `state`.

My goal here was to abstract out the minimal set of metadata required for a compaction service to be able to perform Kafka compaction on a partition. This is generally described in the individual commits. High level components in this PR include:
- the `metastore` interface definition for compaction metadata
- the `state` definition for the data structures that will serve the interface
- the `state_updates` to validate changes to the `state`
- the implementation of `simple_metastore` using `state` and `state_updates`

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

rockwotj commented:
Sorry for the conflicts :(

rockwotj commented on src/v/cloud_topics/level_one/metastore.h:
```diff
@@ -109,6 +110,96 @@ class metastore {
+            bool operator==(const cleaned_range&) const = default;
+            kafka::offset base_offset;
+            kafka::offset last_offset;
+            std::optional<model::timestamp> cleaned_with_tombstones_at;
```
I'm assuming that if the `std::optional` is set then it means that there is both tombstones and this is the timestamp that compaction finished? I do wonder if from an API prospective, if it's easier to understand if we always pass in a timestamp and a separate bool for if there were tombstones, then the impl just ignores the timestamp otherwise. I'm fine to keep the `std::optional`, but let's add a quick note so folks don't have to guess or go read more code.

andrwng replied:
Good point. Also realized that the cleaned timestamp can/should probably be a single timestamp per compaction operation, so I've tweaked this a bit to reflect that, and separated out a bool.

andrwng commented:
> Sorry for the conflicts :(

Ah no problem, let me rebase before folks get too far reviewing this.

andrwng committed:
ct/l1: introduce compaction interfaces to metastore

This commit adds interfaces for updating and getting compaction-related
metadata. In particular, it exposes:
- compact_objects(), which is like replace_objects() but with additional
  metadata about newly cleaned ranges
- get_compaction_offsets(), which tells callers about dirty ranges and
  regions of the log eligible for tombstone removal

This is just an interface. Following commits will add relevent state to
implement it in the simple_metastore.

rockwotj commented on src/v/cloud_topics/level_one/metastore.h:
```diff
@@ -109,6 +110,96 @@ class metastore {
+        offset_interval_set removable_tombstone_ranges;
+    };
+    // Similar to replace_objects(), but with additional constraints based on
+    // compaction metadata.
```
maybe a note to see `get_compaction_offsets` for more information, or place this method after `get_compaction_offsets` so they have to scroll by?

rockwotj commented on src/v/cloud_topics/level_one/metastore.h:
```diff
@@ -109,6 +110,96 @@ class metastore {
+    //   m.max_offset(), m, removed_tombstones_ranges)
+    //
+    // co_await metastore.compact_objects( \
+    //   objects, {new_cleaned_ranges, removed_tombstones_ranges})
```
nit: you changed this to be a `map_t` of partitions, would be good to show how that factors in.

rockwotj commented on src/v/cloud_topics/level_one/state.h:
```diff
@@ -50,6 +51,92 @@ struct extent
+    // avoid complicating reasoning about cleaned ranges. Ordered, and
+    // maintained to be non-overlapping.
+    //
+    // These may overlap with `cleaned_ranges`.
```
These should always be contained by `cleaned_ranges` right?

rockwotj commented on src/v/cloud_topics/level_one/state.h:
```diff
@@ -50,6 +51,92 @@ struct extent
+
+    // Cleaned offset ranges that contain tombstones, tracked separately to
+    // avoid complicating reasoning about cleaned ranges. Ordered, and
+    // maintained to be non-overlapping.
```
I guess we need a `offset_interval_map` for this...

andrwng replied:
Yeah it would be nice, though it's a bit complicated because we need to worry about merging the state in the intervals in that case... I'm punting on this if you don't feel strongly about it

rockwotj replied:
Yeah SGTM

rockwotj commented on src/v/cloud_topics/level_one/state.h:
```diff
@@ -77,15 +166,9 @@ struct partition_state
-    // - When adding objects from compaction, expand the cleaned range.
+    // Describes the current state of compacted regions of this partition's
+    // log, if it belongs to a compacted topic.
+    std::optional<compaction_state> compaction_state;
```
Is this always present for compacted topics or only once compaction runs?

andrwng replied:
Only once compaction runs. I left a TODO to think about this a bit more when toggling cleanup policies, but I don't think it's necessarily harmful to leave it be.

rockwotj commented on src/v/cloud_topics/level_one/state_update.h:
```diff
@@ -81,6 +81,35 @@ struct add_objects_update
+    std::optional<cleaned_range> new_cleaned_range;
+
+    // Shouldn't have a cleaned_with_tombstones_at. Should not overlap.
+    chunked_vector<struct cleaned_range> removed_tombstones_ranges;
```
does this need the `struct` prefix?

andrwng replied:
Nope, leftover. Removed

rockwotj commented on src/v/cloud_topics/level_one/state_update.h:
```diff
@@ -81,6 +81,35 @@ struct add_objects_update
+
+        kafka::offset base_offset;
+        kafka::offset last_offset;
+        std::optional<model::timestamp> cleaned_with_tombstones_at;
```
nit: light docs or at least point to the state object for docs?

andrwng replied:
Changed to a has_tombstones bool.

andrwng committed:
ct/l1: add state for compaction

Adds state that will be used for the simple_metastore's implementation
of compaction metadata.

andrwng committed:
ct/l1: state update for compaction

Adds compaction metadata to the replace_objects state update, allowing
callers to track extra compaction state, and add additional
compaction-specific constraints to object replacement.

Note, the end user (compaction) isn't expected to use these
state_updates directly, and will instead go through the
simple_metastore, which will use this new udpate capability in a later
commit.

andrwng committed:
ct/l1: simple_metastore compaction metadata

Adds an implementation of the compaction-related metastore interface for
simple_metastore, backed by the new state and state_updates.

As is true generally with the simple_metastore, nothing here is
persisted at this point, making the simple_metastore ideal for unit
testing.

andrwng commented:
Haven't addressed feedback, just rebased

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--b4d9f4a5-ad8a-4316-b93b-6fcc895c1c35-->
<!--TEST RESULTS-->
<!--b6b2c6c5-311f-42df-921c-d8ff2bceb518-->
<!--TEST RESULTS-->
<!--6547f846-014b-404f-905d-5c5bec54321c-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69374">build#69374</a></summary>


| test_class            | test_method                   | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------------|:------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| FeaturesMultiNodeTest | test_license_upload_and_query |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69374#01982e90-f5af-4a77-b953-480ad8d31133 | FLAKY         | 19/21    | upstream reliability is '98.48101265822785'. current run reliability is '90.47619047619048'. drift is 8.00482 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69388">build#69388</a></summary>


| test_class                | test_method                   | test_arguments                                                                                                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:--------------------------|:------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest | test_write                    |                                                                                                                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69388#01982f46-995d-4cd6-8b52-6bebaa98b383 | FLAKY         | 18/21    | upstream reliability is '96.61016949152543'. current run reliability is '85.71428571428571'. drift is 10.89588 and the allowed drift is set to 50. The test should PASS |
| FeaturesMultiNodeTest     | test_license_upload_and_query |                                                                                                                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69388#01982f46-995c-40af-885f-9f7e310ac977 | FLAKY         | 18/21    | upstream reliability is '98.48484848484848'. current run reliability is '85.71428571428571'. drift is 12.77056 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest  | test_node_operations          | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69388#01982f46-9958-459f-8a51-baaaa0f216f0 | FLAKY         | 20/21    | upstream reliability is '99.19571045576407'. current run reliability is '95.23809523809523'. drift is 3.95762 and the allowed drift is set to 50. The test should PASS  |
| TestTieredStoragePause    | test_safe_pause_resume        | {"allow_gaps_cluster_level": true, "allow_gaps_topic_level": true}                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69388#01982f46-995b-422c-8c91-6fb90595751d | FLAKY         | 20/21    | upstream reliability is '98.88059701492537'. current run reliability is '95.23809523809523'. drift is 3.6425 and the allowed drift is set to 50. The test should PASS   |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69553">build#69553</a></summary>


| test_class                | test_method                   | test_arguments                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:--------------------------|:------------------------------|:-------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest | test_write                    | null                           | integration | https://buildkite.com/redpanda/redpanda/builds/69553#0198387a-cd84-465d-8103-0a76005eec29 | FLAKY         | 19/21    | upstream reliability is '95.48872180451127'. current run reliability is '90.47619047619048'. drift is 5.01253 and the allowed drift is set to 50. The test should PASS  |
| JavaCompressionTest       | test_upgrade_java_compression | {"compression_type": "gzip"}   | integration | https://buildkite.com/redpanda/redpanda/builds/69553#01983874-f82c-4fc7-82f9-3be13a8ca400 | FLAKY         | 20/21    | upstream reliability is '95.1048951048951'. current run reliability is '95.23809523809523'. drift is -0.1332 and the allowed drift is set to 50. The test should PASS   |
| JavaCompressionTest       | test_upgrade_java_compression | {"compression_type": "gzip"}   | integration | https://buildkite.com/redpanda/redpanda/builds/69553#0198387a-cd82-4bfe-84ef-a5e507e835ba | FLAKY         | 20/21    | upstream reliability is '95.1048951048951'. current run reliability is '95.23809523809523'. drift is -0.1332 and the allowed drift is set to 50. The test should PASS   |
| JavaCompressionTest       | test_upgrade_java_compression | {"compression_type": "lz4"}    | integration | https://buildkite.com/redpanda/redpanda/builds/69553#01983874-f82c-4429-82cc-4a75c2df8520 | FLAKY         | 20/21    | upstream reliability is '92.41379310344827'. current run reliability is '95.23809523809523'. drift is -2.8243 and the allowed drift is set to 50. The test should PASS  |
| JavaCompressionTest       | test_upgrade_java_compression | {"compression_type": "lz4"}    | integration | https://buildkite.com/redpanda/redpanda/builds/69553#0198387a-cd82-4b0d-84f6-9f2e76de01e3 | FLAKY         | 20/21    | upstream reliability is '92.41379310344827'. current run reliability is '95.23809523809523'. drift is -2.8243 and the allowed drift is set to 50. The test should PASS  |
| JavaCompressionTest       | test_upgrade_java_compression | {"compression_type": "snappy"} | integration | https://buildkite.com/redpanda/redpanda/builds/69553#01983874-f82d-428e-9869-d72e32de3e13 | FLAKY         | 20/21    | upstream reliability is '92.46575342465754'. current run reliability is '95.23809523809523'. drift is -2.77234 and the allowed drift is set to 50. The test should PASS |
| JavaCompressionTest       | test_upgrade_java_compression | {"compression_type": "snappy"} | integration | https://buildkite.com/redpanda/redpanda/builds/69553#0198387a-cd83-482a-bd85-0977ccbdce72 | FLAKY         | 20/21    | upstream reliability is '92.46575342465754'. current run reliability is '95.23809523809523'. drift is -2.77234 and the allowed drift is set to 50. The test should PASS |
| JavaCompressionTest       | test_upgrade_java_compression | {"compression_type": "zstd"}   | integration | https://buildkite.com/redpanda/redpanda/builds/69553#01983874-f82f-44d1-bf60-ce002339c7ae | FLAKY         | 20/21    | upstream reliability is '92.3076923076923'. current run reliability is '95.23809523809523'. drift is -2.9304 and the allowed drift is set to 50. The test should PASS   |
| JavaCompressionTest       | test_upgrade_java_compression | {"compression_type": "zstd"}   | integration | https://buildkite.com/redpanda/redpanda/builds/69553#0198387a-cd84-465d-8103-0a76005eec29 | FLAKY         | 20/21    | upstream reliability is '92.3076923076923'. current run reliability is '95.23809523809523'. drift is -2.9304 and the allowed drift is set to 50. The test should PASS   |
</details>
<!--/body-->

andrwng commented:
Force push https://github.com/redpanda-data/redpanda/compare/d82d7abd7c0c7bf4488f1737669da485788c9f10..e0cca2fa5f3221073d66dd0ac7fb0f5a9864f3c4
* Added/tweaked a few comments
* Changed from using std::optional<model::timestamp> to using a model::timestamp and bool for tracking tombstone tracking. I also realized that when we get around to adding metadata for scheduling, knowing when we last cleaned the log will be useful, so we might as well always track this cleaned_at timestamp.

WillemKauf commented on src/v/cloud_topics/level_one/metastore/state.cc:
```diff
@@ -28,10 +28,152 @@ state::partition_state(const model::topic_id_partition& tidp) const {
     return prt_iter->second;
 }
 
+bool compaction_state::has_contiguous_range_with_tombstones(
```
Do we need this? given that `get_contiguous_range_with_tombstones()` isn't cheap, probably better to force a user to call it directly instead of giving the option to write code like

```cpp
if(state.has_contiguous_range_with_tombstones()){
  auto range = state.get_contiguous_range_with_tombstones();
}
```

andrwng replied:
I defined it as public interface and made get_contiguous_range_with_tombstones private to discourage this. But I can make this change if you think it is still prone to confusion as is

andrwng replied:
We discussed on Slack, keeping this as is since the public/private distinction helps avoid the problem

WillemKauf commented on src/v/cloud_topics/level_one/metastore/state.h:
```diff
@@ -50,6 +51,92 @@ struct extent
+
+    // Removes the input inclusive range from the set of cleaned ranges with
+    // tombstones. The input range doesn't need to align exactly with any of
+    // `cleaned_ranges_with_tombstones`, but it must be fully covered.
```
> but it must be fully covered.

Read this a few times over and still wasn't 100% sure what this implies (to me, reads as an ambiguous "it"), can you add a diagram? :sweat_smile:

andrwng replied:
Will do

WillemKauf commented on src/v/cloud_topics/level_one/metastore/state_update.cc:
```diff
@@ -191,6 +191,104 @@ replace_objects_update::can_apply(const state& state) {
+                if (req_cleaned_range.last_offset > req_last) {
+                    return std::unexpected(stm_update_error(fmt::format(
+                      "Cleaned range for {} does not match requested new "
+                      "extents' "
```
weird formatting here?

andrwng replied:
Done

WillemKauf commented on src/v/cloud_topics/level_one/metastore/state_update.h:
```diff
@@ -81,6 +81,42 @@ struct add_objects_update
+      compaction_state_update,
+      serde::version<0>,
+      serde::compat_version<0>> {
+    struct cleaned_range
```
second duplicated definition of `cleaned_range`: should this be defined once elsewhere?

andrwng replied:
I'll leave a comment about this. The duplication is intentional to decouple the metastore interface from underlying implementation/storage details. As the underlying implementation of the state changes, I don't want callers of the metastore to have to change their code, and vice versa.

WillemKauf commented on src/v/cloud_topics/level_one/metastore/simple_metastore.cc:
```diff
@@ -123,4 +123,98 @@ simple_metastore::get_first_ge(
+    const auto& cmp_state = *prt.compaction_state;
+    auto offsets_stream = cmp_state.cleaned_ranges.make_stream();
+    auto dirty_base_candidate = prt.start_offset;
+    while (offsets_stream.has_next()) {
```
Going back to our earlier conversations, the size of this stays bounded because `cleaned_ranges` will merge adjacent ranges, right?

andrwng replied:
That's right

WillemKauf reviewed (commented):
Sorry for the wait, some silly questions/nits on my end

rockwotj commented on src/v/cloud_topics/level_one/metastore/metastore.h:
```diff
@@ -109,6 +110,103 @@ class metastore {
+        // from the start of the log.
+        std::optional<cleaned_range> new_cleaned_range;
+
+        // Ranges of offsets that previously had tombstones, that have been
+        // removed. Each range's `has_tombstones` is assumed to be false and
+        // is ignored.
+        chunked_vector<cleaned_range> removed_tombstones_ranges;
```
should this data structure be `offset_interval_set` then?

andrwng replied:
Ah great point, that does feel more natural. Will make that change.

rockwotj reviewed (dismissed):
LGTM (Sorry I didn't see this - it wasn't in my queue because the re-request review wasn't toggled). Maybe one data structure simplification.

dotnwat commented on src/v/cloud_topics/level_one/metastore/metastore.h:
```diff
@@ -109,6 +110,102 @@ class metastore {
+    // Returns metadata required to determine what to compact for the given
+    // partition. Below is pseudocode for sample usage:
+    //
+    // offsets = co_await metastore.get_compaction_offsets( \
+    //   partition, tombstone_removal_upper_bound_ts);
+    //
+    // key_offset_map m;
+    // cleaned_ranges new_cleaned_ranges;
+    // offset_interval_set removed_tombstones_ranges;
+    //
+    // # Build an offset map based on the dirty ranges.
+    // for dirty_range in offsets.dirty_ranges:
+    //     reader = log.reader(dirty_range.base, dirty_range.last)
+    //     co_await m.add_latest_offset_per_key(reader)
+    //
+    //     cleaned_range r(...offset range that was actually read...);
+    //     if ...reader witnessed tombstones...:
+    //         cleaned_range.cleaned_with_tombstones_at = now()
+    //
+    //     new_cleaned_ranges.insert(cleaned_range)
+    //
+    // # Determine what ranges to remove tombstones from.
+    // removed_tombstones_ranges = \
+    //   ...offsets.removable_tombstone_ranges that fall below m.max_offset()...
+    //
+    // # This operation deduplicates based on the offset map and removes
+    // # tombstones in the given ranges, up to the max indexed by the map.
+    // objects = co_await log.compact( \
+    //   m.max_offset(), m, removed_tombstones_ranges)
+    //
+    // co_await metastore.compact_objects( \
+    //   objects, {{tp, {new_cleaned_ranges, removed_tombstones_ranges}}})
```
🔥

dotnwat reviewed (commented):
so good
